### PR TITLE
Main 17620: Extend jsonenums to support existing iota types

### PR DESCRIPTION
**Pull request by BenjaminTrapani**
AC:
- Use the raw iota string name during db serialization and deserialization instead of the json representation.
- Add a 'prefix' flag that allows the caller to specify a prefix of each iota var name to drop when converting to serialized representation (useful for many of our iota consts that are prefixed with the type name)
- Add a 'all-caps' flag that serializes to a string equal to the var name converted to uppercase and then transformed per the rest of the flags and logic. If converting to all-caps results in collisions (multiple constants serialize to the same value), error.
- Apply jsonenums using these new flags to a few test iota types to demonstrate the functionality.

https://clypdinc.atlassian.net/browse/MAIN-17620

### CODE REVIEW

File: jsonenums.go [hunk 1]
```diff
@@ -67,7 +67,9 @@ package main
```
Comment (quintenpalmer) [inline jsonenums.go:67]: This is a bit hard to read with all of the changes in one commit, but I think this looks good.

File: template.go [hunk 2]
```diff
@@ -75,16 +87,15 @@ func (r *{{$typename}}) Scan(i interface{}) error {
 	case []byte:
 		return r.UnmarshalJSON(t)
```
Comment (quintenpalmer) [inline template.go]: Should this UnmarshalJSON call also be a setValue call instead now? Casting the bytes to a string

**Comment by quintenpalmer** [comment]
@BenjaminTrapani one question about unmarshaling in the `Scan`

In the future I advise splitting up the changes into logical changes per commit so that the reviewer has an easier time understanding each change individually.

I also feel that with the changes we're making to jsonenums it would be good to include more tests in this code and pull more out of the main. I'll log a ticket to capture that, or you can if you'd like?

After resolving the `Scan` questions, this seems good to merge though.

File: template.go (hunk 2 above)
Comment (BenjaminTrapani) [reply template.go]: @quintenpalmer Does the type we get from gorp depend on the db column type or the value returned from the value() function? If the first, we still need this clause and should do as you suggested. If not, we should be able to remove this case entirely.

Comment (quintenpalmer) [reply template.go]: I'm not sure what database types, if any, gorp will pass as a `[]byte`. I think we should have both branches unless we never saw anything going through the `[]byte` case, in which case let's remove it.

**Comment by quintenpalmer** [comment]
👍  from me.

File: jsonenums.go [hunk 3]
```diff
@@ -80,10 +82,12 @@ import (
...
 var (
-	typeNames           = flag.String("type", "", "comma-separated list of type names; must be set")
-	outputPrefix        = flag.String("prefix", "", "prefix to be added to the output file")
-	outputSuffix        = flag.String("suffix", "_jsonenums", "suffix to be added to the output file")
-	exportSnakeCaseJSON = flag.Bool("snake_case_json", false, "Map camel case variable names to snake case json?")
+	typeNames              = flag.String("type", "", "comma-separated list of type names; must be set")
+	outputPrefix           = flag.String("prefix", "", "prefix to be added to the output file")
+	outputSuffix           = flag.String("suffix", "_jsonenums", "suffix to be added to the output file")
+	exportSnakeCaseJSON    = flag.Bool("snake_case_json", false, "Map camel case variable names to snake case json?")
+	serializedPrefixToDrop = flag.String("prefix_to_drop", "", "string to drop from beginning of each iota const name when converting to string")
```
Comment (ashultz0) [inline jsonenums.go:89]: are we ever using this to drop something which is not the class name?  I'm wondering if this should just be a boolean

**Comment by ashultz0** [comment]
one flag question

File: jsonenums.go (hunk 3 above)
Comment (BenjaminTrapani) [reply jsonenums.go:89]: @ashultz0 There are cases where the two do not match exactly, see InventoryClassType.

**Comment by ashultz0** [comment]
👍

**Pull request merged**